src/Rings/MPolyMap/flattenings.jl

```diff
@@ -0,0 +1,247 @@
+
+  return S_flat
+end
+
+@attr function _flatten(
+    S::AbstractAlgebra.Generic.MPolyRing{RingElemType}
+  ) where {RingElemType <: MPolyLocalizedRingElem}
+  L = base_ring(S)::MPolyLocalizedRing
+  R = base_ring(L)::MPolyRing
+  U = inverted_set(L)::AbsMultSet
+  kk = coefficient_ring(R)::Field
+
+  # Before building S_flat, we have to create a polynomial 
+  # ring T_flat from which we can pass to the localized quotient.
+  T_flat, _ = PolynomialRing(kk, vcat(symbols(S), symbols(R)))
+  R_to_T_flat = hom(R, T_flat, gens(T_flat)[ngens(S)+1:end])
+
+  S_flat, _ = localization(T_flat, R_to_T_flat(U)) # Will throw if multiplicative set can not be transfered. 
+  L_to_S_flat = hom(L, S_flat, gens(S_flat)[ngens(S)+1:end])
+  
+  R_to_S_flat = hom(R, S_flat, gens(S_flat)[ngens(S)+1:end])
+  
+  S_to_S_flat = hom(S, S_flat, L_to_S_flat, gens(S_flat)[1:ngens(S)])
+  S_flat_to_S = hom(S_flat, S, vcat(gens(S), S.(gens(R))))
+
+  set_attribute!(S, :_flat_to_orig, S_flat_to_S)
+  set_attribute!(S, :_orig_to_flat, S_to_S_flat)
+  set_attribute!(S, :_base_ring_to_flat, L_to_S_flat)
+
+  return S_flat
+end
+
+### Transfer of multiplicative sets along ring homomorphisms 
+function (phi::MPolyAnyMap{<:MPolyRing, <:MPolyRing, Nothing})(U::MPolyPowersOfElement;
+                                                               check::Bool=true
+                                                              )
+  ambient_ring(U) === domain(phi) || error("multiplicative set does not lay in the domain of the morphism")
+  S = codomain(phi) 
+  SU = MPolyPowersOfElement(S, phi.(denominators(U)))
+  return SU
+end
+
+function (phi::MPolyAnyMap{<:MPolyRing, <:MPolyRing, Nothing})(U::MPolyComplementOfPrimeIdeal;
+                                                               check::Bool=true
+                                                              )
+  ambient_ring(U) === domain(phi) || error("multiplicative set does not lay in the domain of the morphism")
+  S = codomain(phi) 
+  Q = ideal(S, phi.(gens(prime_ideal(U))))
+  SU = MPolyComplementOfPrimeIdeal(S, Q, check=check)
+  return SU
+end
+
+### Canonical maps involved in flattening of towers of polynomial rings
+@attr function _flat_to_orig(
+    S::AbstractAlgebra.Generic.MPolyRing{RingElemType}
+  ) where {RingElemType <: Union{<:MPolyElem, <:MPolyQuoElem, <:MPolyLocalizedRingElem, 
+                                 <:MPolyQuoLocalizedRingElem
+                                }
+          }
+  _flatten(S) # Will set the attribute
+  return get_attribute(S, :_flat_to_orig)
+end
+
+@attr function _orig_to_flat(
+    S::AbstractAlgebra.Generic.MPolyRing{RingElemType}
+  ) where {RingElemType <: Union{<:MPolyElem, <:MPolyQuoElem, <:MPolyLocalizedRingElem, 
+                                 <:MPolyQuoLocalizedRingElem
+                                }
+          }
+  _flatten(S) # Will set the attribute
+  return get_attribute(S, :_orig_to_flat)
+end
+
+@attr function _base_ring_to_flat(
```

Comment:
Should not have the `@attr`.

Reply:
The `@attr` is weird in the sense that the body of the function does not really produce the attribute to be cached, but it calls another function which manually caches it. But I didn't find a better way to do this. The map is created together with the flattened ring and, hence, should already be cached there. Moreover, it does not make sense to reproduce this map every time it is being called. 

Another solution would be to cache everything in the attributes of the flattened ring, remove the `@attr` here and write the body as a getter to the respective attribute. Better?

Reply:
I don't see how the `@attr` adds anything. The `_flatten(S)` call is already cached and the return type is not used. So it is basically a noop. And the `@attr` here just does the `get_attribute`, which is already done in the body of the function.

This can just be written as
```
function _base_ring_to_flat(S)
  _flatten(S) # Will set the attribute or do nothing it if is set
  return get_attribute(S, :_base_ring_to_flat)
end
```